lock-api/src/main/java/com/palantir/lock/client/timestampleases/MultiClientMinLeasedTimestampGetter.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client.timestampleases;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.atlasdb.autobatch.Autobatchers;
+import com.palantir.atlasdb.autobatch.DisruptorAutobatcher;
+import com.palantir.atlasdb.futures.AtlasFutures;
+import com.palantir.atlasdb.timelock.api.Namespace;
+import com.palantir.atlasdb.timelock.api.TimestampLeaseName;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.lock.client.InternalMultiClientConjureTimelockService;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Set;
+
+public final class MultiClientMinLeasedTimestampGetter implements AutoCloseable {
+    private final DisruptorAutobatcher<MinLeasedTimestampGetterBatchParams, Long> disruptor;
+
+    private MultiClientMinLeasedTimestampGetter(
+            DisruptorAutobatcher<MinLeasedTimestampGetterBatchParams, Long> disruptor) {
+        this.disruptor = disruptor;
+    }
+
+    static MultiClientMinLeasedTimestampGetter create(InternalMultiClientConjureTimelockService delegate) {
+        DisruptorAutobatcher<MinLeasedTimestampGetterBatchParams, Long> disruptor = Autobatchers.coalescing(
+                        new MinLeasedTimestampGetterBatchHandler(delegate))
+                .safeLoggablePurpose("min-leased-timestamp-getter")
+                .batchFunctionTimeout(Duration.ofSeconds(30))
+                .build();
+
+        return new MultiClientMinLeasedTimestampGetter(disruptor);
+    }
+
+    public Map<TimestampLeaseName, Long> getMinLeasedTimestamps(
+            Namespace namespace, Set<TimestampLeaseName> timestampNames) {
+        Map<TimestampLeaseName, ListenableFuture<Long>> futures = KeyedStream.of(timestampNames)
+                .map(timestampName -> MinLeasedTimestampGetterBatchParams.of(namespace, timestampName))
+                .map(this::getMinLeasedTimestamp)
+                .collectToMap();
+
+        return AtlasFutures.getUnchecked(AtlasFutures.allAsMap(futures, MoreExecutors.directExecutor()));
+    }
```

Comment:
In the `acquireTimestampLeases`, I have the autobatcher's batch element match `Map<TimestampName, Long>`, which is the input that will be given by `TimelockService`.

The reason for that is the autobatcher is an "independent" one and as such each request needs to be fulfilled on its own regard. I assumed it's better to keep the batch unit somewhat large (to avoid partial failures).

With `getMinLeasedTimestamps`, I can use the coalescing batcher. For any given timestamp name and namespace, only one request is needed and so partial failure is less likely.

Let me know if any of this makes sense. And, any thoughts on what the batch unit should be!

Reply:
What failures are we talking about? We should only have a single outstanding request here, I feel